ciao-controller/main.go

```diff
@@ -273,7 +289,23 @@ func (c *controller) startCiaoService() error {
 
 	service := fmt.Sprintf(":%d", controllerAPIPort)
 
-	glog.Infof("Starting ciao API on port %d\n", controllerAPIPort)
+	glog.Infof("Starting ciao API on port %d", controllerAPIPort)
+
+	server := &http.Server{
+		Handler: r,
+		Addr:    service,
+	}
 
-	return http.ListenAndServeTLS(service, httpsCAcert, httpsKey, r)
+	c.httpServers = append(c.httpServers, server)
+	return server.ListenAndServeTLS(httpsCAcert, httpsKey)
+}
+
+func (c *controller) ShutdownHTTPServers() {
+	glog.Warning("Shutting down HTTP servers")
+	for _, server := range c.httpServers {
+		err := server.Shutdown(context.TODO())
```

Comment:
We should use a timeout context here.  There's no guarantee that shutdown will ever return.  We could call shutdown in parallel with a time out of say 1 second.